src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,172 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private static final double HALF = 0.5;
+    private static final int ONE = 1;
+    private int size;
+    private Node<T> head;
+    private Node<T> tail;
+
+    private static class Node<T> {
+        private Node<T> next;
+        private Node<T> prev;
+        private T value;
+
+        Node(T value) {
+            this.value = value;
+            this.next = null;
+            this.prev = null;
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value);
+        if (isEmpty()) {
+            head = newNode;
+        } else {
+            tail.next = newNode;
+            newNode.prev = tail;
+        }
+        tail = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        validateIndexNotInclusive(index);
+
+        if (index == size) {
+            add(value);
+            return;
+        }
+
+        Node<T> newNode = new Node<>(value);
+        Node<T> currentNode = head;
+
+        if (index == 0) {
+            newNode.next = head;
+            if (head != null) {
+                head.prev = newNode;
+            }
+            head = newNode;
+            size++;
+            return;
+        }
+
+        currentNode = findNodeByIndex(index - ONE);
+        Node<T> nextNode = currentNode.next;
+        newNode.next = nextNode;
+        if (nextNode != null) {
+            nextNode.prev = newNode;
+        }
+        currentNode.next = newNode;
+        newNode.prev = currentNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
```

Comment:
```suggestion
    public void addAll(List<T> list) {
        for (T element : list) {
            add(element);
        }
    }
```
looks more readable in my opinion